src/io/flutter/run/daemon/FlutterAppManager.java

```diff
@@ -346,9 +364,11 @@ private void error(JsonObject json) {
   }
 
   private RunningFlutterApp findApp(FlutterDaemonController controller, String appId) {
-    for (FlutterApp app : myApps) {
-      if (app.getController() == controller && app.hasAppId() && app.appId().equals(appId)) {
-        return (RunningFlutterApp)app;
+    synchronized (myLock) {
+      for (FlutterApp app : myApps) {
+        if (app.getController() == controller && app.hasAppId() && app.appId().equals(appId)) {
```

Comment:
The very very lastest flutter codebase returns unique app ids as well, even across daemon sessions. Not sure if you want to update the code to reflect that yet though, since it depends on everybody being in the most up to date flutter.